### PR TITLE
Ensure conditional validation has knowledge of old and new object

### PR DESCRIPTION
**Pull request by liggitt**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
Moves conditional validation to methods with knowledge of the old and new object, to ensure we never fail an object based on a validation check of data which is already in the existing persisted object.

This ensures that once the SCTP feature is enabled, rollback of one version is safe (API validation-wise).

**Which issue(s) this PR fixes**:
Finishes https://github.com/kubernetes/kubernetes/issues/72651

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

/sig api-machinery
/cc @sbezverk

### CODE REVIEW

File: pkg/apis/core/validation/conditional_validation.go [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+		for pIndex, p := range s.Ports {
+			if p.Protocol == api.ProtocolSCTP {
+				fields = append(fields, field.NewPath("subsets").Index(sIndex).Child("ports").Index(pIndex).Child("protocol"))
+			}
+		}
+	}
+	return fields
+}
+
+// ValidateConditionalPodTemplate validates conditionally valid fields.
```
Comment (sbezverk) [inline pkg/apis/core/validation/conditional_validation.go]: I could not find Unit tests for `ValidateConditionalPodTemplate` any particular reason it does not need one?

**Review by sbezverk — COMMENTED** [review]
I need just 1 clarification, other than that looks good.

File: pkg/apis/core/validation/conditional_validation.go (hunk 1 above)
Comment (liggitt) [reply pkg/apis/core/validation/conditional_validation.go]: both `ValidateConditionalPodTemplate` and `ValidateConditionalPod` route to `validateConditionalPodSpec`, where the validation logic lives, so I just added tests for the latter.

Comment (sbezverk) [reply pkg/apis/core/validation/conditional_validation.go]: thanks

**Pull request merged**